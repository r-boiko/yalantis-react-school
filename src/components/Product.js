// Core
import React from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import PropTypes from 'prop-types';

// Antd
import {Card, Icon} from "antd";

// Components
import UpdateProduct from "./UpdateProduct";

// Store
import {addProductsToCart, removeProductsFromCart} from "../store/cart/actions";

// Selectors
import {getCart} from "../store/selectors";

// PropTypes
Product.propTypes = {
  single: PropTypes.bool,
  id: PropTypes.string,
  name: PropTypes.string,
  price: PropTypes.number,
  origin: PropTypes.string,
  createdAt: PropTypes.string,
  updatedAt: PropTypes.string,
  isEditable: PropTypes.bool,
  cart: PropTypes.object,
  addProductsToCart: PropTypes.func,
  removeProductsFromCart: PropTypes.func,
};

function Product({single, id, name, price, origin, createdAt, updatedAt, isEditable, addProductsToCart, removeProductsFromCart, cart}) {
  const productFromCart = cart.items.find(item => item.id === id);

  const actions = [
    !isEditable ? (<span>
            <Icon
                key="shopping-cart"
                type="shopping-cart"
                onClick={() => addProductsToCart({id, name, price, origin, createdAt, updatedAt})}
            />
      {productFromCart && productFromCart.count}
        </span>) : (
        <UpdateProduct id={id} name={name} price={price} origin={origin}/>
    )
    ,
    !isEditable && (
        <Icon
            key="delete"
            type="delete"
            onClick={() => removeProductsFromCart({id})}
        />
    )
    ,
    !single && (
        <Link key="star" to={`/products/${id}`}>
          <Icon type="link"/>
        </Link>
    )
  ].filter(Boolean);

  return (
      <Card style={{width: 300, margin: 20}} actions={actions}>
        <p>Name: {name}</p>
        <p>Price: {`${price}$`}</p>
        <p>Origin: {origin}</p>
        <p>Created: {new Date(createdAt).toDateString()}</p>
        <p>Updated: {new Date(updatedAt).toDateString()}</p>
      </Card>
  );
}

const mapStateToProps = (state) => {
  return {
    cart: getCart(state)
  }
};

const mapDispatchToProps = dispatch => {
  return {
    addProductsToCart: product => {
      dispatch(addProductsToCart(product))
    },
    removeProductsFromCart: id => {
      dispatch(removeProductsFromCart(id))
    }
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(Product);