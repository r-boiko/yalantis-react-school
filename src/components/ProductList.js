// Core
import React from "react";
import PropTypes from 'prop-types';

// Components
import Product from "./Product";

// PropTypes
ProductList.propTypes = {
  products: PropTypes.array,
};

export default function ProductList({products}) {
  return (
      <div className="content">
        <div className="products">
          {products.length === 0 ? 'No data' : products.map((product) => <Product key={product.id} {...product} />)}
        </div>
      </div>
  );
}
