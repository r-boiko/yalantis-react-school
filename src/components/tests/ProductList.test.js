// Core
import React from "react";

// Components
import ProductList from "../ProductList";
import Product from "../Product";

// Enzyme
import {shallow} from 'enzyme';

const defaultProps = {
  products: [{
    isEditable: false,
    id: "4052c0d6-8eae-4237-b1be-9e02033c8799",
    name: "Handmade Frozen Pizza",
    price: 439,
    origin: "asia",
    createdAt: "2020-01-31T09:31:06.618Z",
    updatedAt: "2020-01-31T09:31:06.618Z"
  }]
};

describe('ProductList', () => {
  it('Initial', () => {
    const ProductListContainer = shallow(<ProductList {...defaultProps}/>);
    expect(ProductListContainer).toMatchSnapshot();
  });

  it('Render', () => {
    const ProductListContainer = shallow(<ProductList {...defaultProps}/>);
    expect(ProductListContainer.find(Product)).toHaveLength(1);
  });

  it('No data', () => {
    const props = {
      products: []
    };
    const ProductListContainer = shallow(<ProductList {...props}/>);
    expect(ProductListContainer.text()).toEqual('No data');
  })
});