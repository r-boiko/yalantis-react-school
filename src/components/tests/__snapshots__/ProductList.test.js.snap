// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Initial 1`] = `ShallowWrapper {}`;

exports[`ProductList initial 1`] = `ShallowWrapper {}`;
